Kruize Notification prod fix

msvinaykumar opened the pull request:
## Description

Previously, we had too many variables in the set, such as experiment_name, container_name, endtime, level, termname, modelname, code, type, and message, which significantly increased cardinality. Now, we've reduced it to just a few variables, like termname, modelname, and type, which has decreased the cardinality, resulting in no more than 10 entries, whereas before, it generated millions.

there will be one follow up MR .. to unblock kruize_recommendations entry at ros side

A metric entry with only three tags, which are unlikely to have many unique values, will therefore help reduce cardinality.

Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Docs update
- [ ] Breaking change (What changes might users need to make in their application due to this PR?)
- [ ] Requires DB changes

## How has this been tested?


A metric entry with only three tags, which are unlikely to have many unique values, will therefore help reduce cardinality.


- [ ] New Test X
- [ ] Functional testsuite

**Test Configuration**
* Kubernetes clusters tested on: 

## Checklist :dart:

- [ ] Followed coding guidelines
- [ ] Comments added
- [ ] Dependent changes merged
- [ ] Documentation updated
- [ ] Tests added or updated

## Additional information

chandrams commented on src/main/java/com/autotune/metrics/KruizeNotificationCollectionRegistry.java:
```diff
@@ -78,7 +82,10 @@ public void createCounterTag(String level, String term, String model, Collection
+                String kibanaLog =  String.format(KruizeConstants.KRUIZE_RECOMMENDATION_METRICS.notification_format_for_KIBANA, this.experiment_name, this.container_name, KruizeConstants.DateFormats.simpleDateFormatForUTC.format(this.interval_end_time), level, term, model, String.valueOf(recommendationNotification.getCode()), recommendationNotification.getType(), recommendationNotification.getMes ...
+                String metricEntry =  String.format(KruizeConstants.KRUIZE_RECOMMENDATION_METRICS.notification_format_for_METRICS,  term, model, recommendationNotification.getType());
+                LOGGER.info(kibanaLog);
+                additionalTags = additionalTags.and(KruizeConstants.KRUIZE_RECOMMENDATION_METRICS.TAG_NAME,metricEntry); // A metric entry with only three tags, which are unlikely to have many unique values, will therefore help reduce cardinality.
```
@msvinaykumar - Can you update on how this has been tested?

chandrams replied:
Has this been tested with cloud watch?

msvinaykumar replied:
No

msvinaykumar commented:
@kusumachalasani  for cloud watch is current production configured to include  LOGGER.INFO  ?

kusumachalasani commented:
> @kusumachalasani for cloud watch is current production configured to include LOGGER.INFO ?

Yes

msvinaykumar commented:
![image](https://github.com/user-attachments/assets/77f65fd9-92e3-44cc-a44e-3e691a4af88e)
![image](https://github.com/user-attachments/assets/474f8991-3273-4297-9cc1-f17e14d452b2)


metrics and cloud watch tested and attached screenshot

msvinaykumar commented:
Cloud Watch
  2024-08-3020:28:46.188 INFO [qtp844112759-63][KruizeNotificationCollectionRegistry.java(87)]-quarkus-resteasy-kruize-min-http-response-time-db_0|tfb-server-0|2024-09-01T01:58:43.302Z|model|short_term|cost|221001|error|Number of pods cannot be zero


metrics

# TYPE KruizeNotifications_total counter
KruizeNotifications_total{api="updateRecommendations",application="Kruize",recommendations_notifications="short_term|cost|error",} 6.0
KruizeNotifications_total{api="updateRecommendations",application="Kruize",recommendations_notifications="short_term|performance|error",} 6.0

kusumachalasani commented:
> # TYPE KruizeNotifications_total counter
> KruizeNotifications_total{api="updateRecommendations",application="Kruize",recommendations_notifications="short_term|cost|error",} 6.0 KruizeNotifications_total{api="updateRecommendations",application="Kruize",recommendations_notifications="short_term|performance|error",} 6.0

@msvinaykumar ? I understand we want to avoid the experiment_name as well in the notification metrics because of cardinality issues, but  how can these metrics help the customer ? As we are sending this info in the logs and the user has an option to look into the logs for debugging purposes. Unless there is an importance for these metrics, it is going to affect the performance.

msvinaykumar commented on src/main/java/com/autotune/metrics/KruizeNotificationCollectionRegistry.java:
```diff
@@ -78,7 +82,10 @@ public void createCounterTag(String level, String term, String model, Collection
             Tags additionalTags = Tags.empty();
             if (("|" + KruizeDeploymentInfo.log_recommendation_metrics_level + "|").contains("|" + recommendationNotification.getType() + "|") == true) {
-                additionalTags = additionalTags.and(KruizeConstants.KRUIZE_RECOMMENDATION_METRICS.TAG_NAME, String.format(KruizeConstants.KRUIZE_RECOMMENDATION_METRICS.notification_format, this.experiment_name, this.container_name, KruizeConstants.DateFormats.simpleDateFormatForUTC.format(this.interval_end_time), level, term, model, String.valueOf(recommendationNotification.getCode()), recommenda ...
+                String kibanaLog =  String.format(KruizeConstants.KRUIZE_RECOMMENDATION_METRICS.notification_format_for_KIBANA, this.experiment_name, this.container_name, KruizeConstants.DateFormats.simpleDateFormatForUTC.format(this.interval_end_time), level, term, model, String.valueOf(recommendationNotification.getCode()), recommendationNotification.getType(), recommendationNotification.getMes ...
```
Done

msvinaykumar commented on src/main/java/com/autotune/utils/KruizeConstants.java:
```diff
@@ -697,8 +697,9 @@ public static final class KRUIZE_CONFIG_DEFAULT_VALUE {
-        public static final String TAG_NAME = "experiment_details";
-        public static final String notification_format = "%s|%s|%s|%s|%s|%s|%s|%s|%s"; //experiment_name,container_name,endtime,level,termname,modelname,code,type,message
+        public static final String TAG_NAME = "recommendations_notifications";
+        public static final String notification_format_for_KIBANA = "%s|%s|%s|%s|%s|%s|%s|%s|%s"; //experiment_name,container_name,endtime,level,termname,modelname,code,type,message
```
Logger.info by default adds timestamp level and filename ... Please refer screenshot and example shown in this PR.

msvinaykumar commented:
> > # TYPE KruizeNotifications_total counter
> > KruizeNotifications_total{api="updateRecommendations",application="Kruize",recommendations_notifications="short_term|cost|error",} 6.0 KruizeNotifications_total{api="updateRecommendations",application="Kruize",recommendations_notifications="short_term|performance|error",} 6.0
> 
> @msvinaykumar ? I understand we want to avoid the experiment_name as well in the notification metrics because of cardinality issues, but how can these metrics help the customer ? As we are sending this info in the logs - the user has an option to look into the logs for debugging purposes. Unless there is an importance for these metrics, it is going to affect the performance.


Ideally, if everything is functioning properly, there shouldn't be any metrics for errors or critical notifications. If such metrics are found, it is a clear indication that the recommendations are compromised and need to be addressed. The SRE or other relevant teams should be able to debug the issue using Kibana logs. These metrics are essential for alerting users to the presence of errors that need to be resolved. Hope it answers your question.

kusumachalasani commented on src/main/java/com/autotune/utils/KruizeConstants.java:
```diff
@@ -697,8 +697,9 @@ public static final class KRUIZE_CONFIG_DEFAULT_VALUE {
-        public static final String TAG_NAME = "experiment_details";
-        public static final String notification_format = "%s|%s|%s|%s|%s|%s|%s|%s|%s"; //experiment_name,container_name,endtime,level,termname,modelname,code,type,message
+        public static final String TAG_NAME = "recommendations_notifications";
+        public static final String notification_format_for_KIBANA = "%s|%s|%s|%s|%s|%s|%s|%s|%s"; //experiment_name,container_name,endtime,level,termname,modelname,code,type,message
```
Can we avoid "kibana" naming in " String notification_format_for_KIBANA" as well ?

msvinaykumar replied:
Done

kusumachalasani commented:
It is important to note how much additional space the logs might occupy due to this change. Considering we run at least 90,000 recommendations daily in the remote monitoring scenario—and there are no request/limit notifications for one-fourth of the experiments—these logs are frequently generated. Understanding this can help us anticipate the extra storage required for a pod and potential errors if the pod runs for an extended period.

kusumachalasani reviewed (approved):
LGTM from a functional point of view. Would be worth measuring the extra storage required if  there is atleast 1 notification for 5 experiments and extrapolate the same for 90K recommendations for production usecase.

msvinaykumar commented:
> LGTM from a functional point of view. Would be worth measuring the extra storage required if there is atleast 1 notification for 5 experiments and extrapolate the same for 90K recommendations for production usecase.

This change doesn't store data in the Postgres database; it simply fetches, processes, and displays the information in cloudwatch and increase the counter at Prometheus  .

chandrams reviewed (approved):
LGTM

msvinaykumar commented:
> > > LGTM from a functional point of view. Would be worth measuring the extra storage required if there is atleast 1 notification for 5 experiments and extrapolate the same for 90K recommendations for production usecase.
> > 
> > 
> > This change doesn't store data in the Postgres database; it simply fetches, processes, and displays the information in cloudwatch and increase the counter at Prometheus .
> 
> I was thinking from the perspective of pod log.

True , Depends on cloudwatch scale limitations.   We have flag to turn off jic

dinogun commented:
@msvinaykumar It appears that there is a build failure and the crc PR check is failing

msvinaykumar commented:
> @msvinaykumar It appears that there is a build failure and the crc PR check is failing

checking

msvinaykumar commented:
> > @msvinaykumar It appears that there is a build failure and the crc PR check is failing
> 
> checking

Thanks, @dinogun, for catching this. I overlooked and missed checking in one file after addressing the review comments. After checking in the file, the CRC passed.
